More accurate outer checks in patterns

retronym opened the pull request:
Avoids eliding outer checks that matter (run/t11534b.scala) and
avoids emitting checks that don't (pos/t11534.scala) which avoids
compiler warnings when the tested class doesn't have an outer
field.

The latter stops the annoying unchecked warning that appeared since
a recent refactoring made `TermName` a final class.

Fixes scala/bug#11534

retronym commented:
Pushed another version to try to walk the tightrope between `pos/t11534` and `run/reflect-priv-ctor.scala`. Marking as WIP as this isn't super well thought through...

eed3si9n commented:
@retronym Would this make 2.12.9?

retronym commented:
@eed3si9n No, I'd rather not rush this one.

sjrd commented:
Could I remind people about this PR? I still have to compile the Scala.js compiler plugin without `-Xfatal-warnings` in the latest 2.12.x nightly.

adriaanm commented:
This PR needs more work before we can merge it. I can't pick it up right now.

dwijnand commented:
> This PR needs more work before we can merge it. I can't pick it up right now.

@adriaanm / @retronym what's remaining and how terrible would it be if this landed as is?  It's looks nice and small, to my amateur eyes.

dwijnand commented:
@retronym could you look at this one, please?  Even just to get more going on improving it.  (I'd like to get rid of the "The outer reference in this type test cannot be checked at run time" exclusion we have in our build.sbt.)

dwijnand commented:
@joroKr21 can you spot where this PR needs more attention before merging?  I'm happy to do the work, I'm just not sure what work there is to do - perhaps you do.

joroKr21 commented:
I'm not sure what is the reasoning here :thinking:

joroKr21 commented:
> Ideally the compiler would be smart enough to avoid the runtime test in this case as we statically know that the prefixes are the same.

I can't convince myself we know that statically.

dwijnand commented on src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala:
```diff
@@ -349,7 +349,17 @@ trait MatchTreeMaking extends MatchCodeGen with Debugging {
+              }
+            case tp => tp.prefix
+          }
+          val testedPrefix = prefixOf(abstractTypesToBounds(testedBinder.info))
```
Good news 🎉  (IMHO) I rebased this on 2.13 (where fixes have been landing over the last 18 months) to study it and found that the whole `RefinedType` shenanigans above (as well as normalizing) isn't required.  The following seems sufficient to get TermName usage to stop warning (i.e. the pos test passes):

```diff
-          val expectedPrefix = expectedTp.prefix
-          val testedPrefix = testedBinder.info.prefix
+          val expectedPrefix   = expectedTp.prefix
+          val testedTp         = testedBinder.info
+          val testedTpToBounds = abstractTypesToBounds(testedTp)
+          val testedPrefix     = testedTpToBounds.prefix
```

While I'm here, here are the values and what subclass they are:

```
expectedTp           = Test1.g.TermName     :        ClassNoArgsTypeRef
expectedPrefix       = Test1.g.type         :          UniqueSingleType
testedBinder         = value x1             :                TermSymbol
testedTp             = sym.NameType         :     AbstractNoArgsTypeRef
testedTpToBounds     = Test1.g.Name         :        ClassNoArgsTypeRef
testedPrefix         = Test1.g.type         :          UniqueSingleType
```

So then it's a question of whether it's sane/principled/theory to throw in a little `abstractTypesToBounds` type mapping here...  Looks like pre-existing usages of that type map are all related to inferring types in patterns (`inferTypedPattern` and `inferConstructorInstance`) which is some form of "precedent".  Collapsed it makes this a 1 line change, too.

joroKr21 replied:
I think the `RefinedType` shenanigans were pushed to satisfy existing tests:
> Pushed another version to try to walk the tightrope between pos/t11534 and run/reflect-priv-ctor.scala. Marking as WIP as this isn't super well thought through...

Did you check those too?

dwijnand replied:
`pos/t11534` is the test I was studying but, you're right, my reduction causes `abstractTypesToBounds` to induce a new outer reference warning, by finding the "wrong" prefix:

```
expectedTp           = reflect.runtime.universe.MethodSymbol                                                                                        :     AbstractNoArgsTypeRef
expectedPrefix       = reflect.runtime.universe.type                                                                                                :          UniqueSingleType
testedBinder         = value p3                                                                                                                     :                TermSymbol
testedTp             = reflect.runtime.universe.MethodSymbol                                                                                        :     AbstractNoArgsTypeRef
straightPrefix       = reflect.runtime.universe.type                                                                                                :          UniqueSingleType
testedTpToBounds     = reflect.runtime.universe.MethodSymbolApi with reflect.runtime.universe.TermSymbolApi with reflect.runtime.universe.SymbolApi :              RefinedType0
testedPrefix         = <notype>                                                                                                                     :                   NoType$
```

Which also reminds me of [this comment](https://github.com/lampepfl/dotty/issues/2156#issuecomment-290701098) in the Dotty repo:

> this makes it sound like there may not need more then one outer-accessor. Here's a counter-example
> 
> ```scala
> class A{ class B }
> class C extends A {self => 
>   class D {
>      final class E extends self.B  // needs two outer-accessors to support all patter-matchings
>   }
> }
> ```

Is that the fix then - to consider multiple prefixes?  The prefix of `sym.Name` is `sym` which isn't `g` but mapping to bounds `g.Name`'s prefix _is_ `g`, while the prefix of `universe.MethodSymbol` is `universe` which is what is expected.  Is that enough?  Is that unsafe?  🤷🏼 🤔 💭

joroKr21 replied:
Yeah I don't know if mapping to the bounds is safe. Who says extending a bound from another prefix is not allowed?

sjrd replied:
Oh, yes, you can do pretty nasty stuff with prefixes in extends clauses. For example:
https://github.com/scala-js/scala-js/blob/465a7645f06325ff9785bb2529c20ebd48ca3df8/test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/NestedJSClassTest.scala#L660-L691

retronym commented:
Thanks for helping to revive this PR! I've spent some time reminding myself with the terrain, here are some notes.

During typechecking of patterns, some checks are performed to rule out statically false tests. Given `def pat(a: A) = a match { case _: B }`, `B` could either be a subtype of `B` or there must be a possibility that a value of some other type that subclasses both `A` and `B` is instantiable. (ie `isPopulated(A, B)` holds.).

When `isPopulated` is used, it ensures the prefixes align.

These analyses need to account for abstract types. In the test in this PR, the abstract type is the type alias `Symbol#NameType`, the return type of `symbol.name`. The motivation for this is to statically return `TermName`/`TypeName` from `(s: TermSymbol) => s.name`, `(s: TermSymbol) => s.name` respectively.

-------------

Considering:

```scala
object Test {
  class O { class A; class B extends A; trait C; class D extends o2.B with C }
  val o1 = new O;
  val o2 = new O;
  def pat1(a: Test.o1.C) = a match { case _: Test.o1.B => true; case _ => false}
  def main(args: Array[String]): Unit = {
    assert(!pat1(new o1.D))
  }
}

```

This gives no unchecked warning, but the assertion doesn't hold. In this case, the `expectedTp` is `o1.C with o1.B` (a `RefinedType`), and `expectedTp.prefix` is `NoType`. This disables outer synthesis in `withOuterTest` (see `if (expectedPrefix eq NoPrefix`...`).

Even if we looked inside the refined type to see that the prefix was `o1.type` in all elements, it would just lead to the _next_ way of disabling the outer check (`testedPrefix =:= expectedPrefix`), which would also be incorrect!

--------------

I have some separate improvements to this area in https://github.com/retronym/scala/pull/109/files. I don't think they overlap with the problem here, but noting here to help me keep track of things.

--------------

A simpler version of the "shallow" fix for the bug might be:

```
          val testedBinderBaseType = testedBinder.info.baseType(testedBinder.info.typeSymbol)
          val testedPrefix = abstractTypesToBounds(testedBinderBaseType).prefix
```

-------------

Sorry to further muddy the waters here! Hopefully the test case above can help us to get to the heart of the issue. 🤞🏽

dwijnand commented:
> This gives no unchecked warning, but the assertion doesn't hold.

Is `isPopulated` assuming a closed world?  I assume it can't and so I don't see why there couldn't be, separately compiled, an alternative `class D2 extends o1.B with C` which would hit that match case.

dwijnand commented:
> A simpler version of the "shallow" fix for the bug might be:
> 
> ```scala
>           val testedBinderBaseType = testedBinder.info.baseType(testedBinder.info.typeSymbol)
>           val testedPrefix = abstractTypesToBounds(testedBinderBaseType).prefix
> ```

Sadly doesn't look like it does, with `run/reflect-priv-ctor.scala`:

```scala
testedTp             = reflect.runtime.universe.MethodSymbol                                                                                         :     AbstractNoArgsTypeRef
testedBinderBaseType = reflect.runtime.universe.MethodSymbol                                                                                         :     AbstractNoArgsTypeRef
testedBounds         = reflect.runtime.universe.MethodSymbolApi with reflect.runtime.universe.TermSymbolApi with reflect.runtime.universe.SymbolApi  :              RefinedType0
testedPrefix         = <notype>                                                                                                                      :                   NoType$
!! 1 - run/reflect-priv-ctor.scala               [output differs]
% diff /d/scala/test/files/run/reflect-priv-ctor.check /d/scala/test/files/run/reflect-priv-ctor-run.log
@@ -1 +1,4 @@
+reflect-priv-ctor.scala:16: warning: The outer reference in this type test cannot be checked at run time.
+    val init  = typeOf[Foo].members find { case m: MethodSymbol => m.isConstructor case _ => false } get
+                                                 ^
```

retronym commented:
@dwijnand Formulating the test this way might make my point a bit clearer:

```scala
object Test {
  case class O(i: Int) {
    class A
    class B extends A {
      def bOuter = O.this
    }
    trait C {
      def cOuter = O.this
    }
    class D extends o2.B with C
  }
  val o1 = new O(1);
  val o2 = new O(2);
  def pat1(a: Test.o1.C) = a match {
    case b: Test.o1.B =>
      assert(b.bOuter eq Test.o1,
             s"expected ${o1} as outer of value conforming to pattern `b: Test.o1.B`, but got ${b.bOuter}")
    case _ =>

  }
  def main(args: Array[String]): Unit = {
    pat1(new o1.D)
  }
}
```

```
qscalac -d /tmp sandbox/test.scala && qscala -cp /tmp Test
java.lang.AssertionError: assertion failed: expected O(1) as outer of value conforming to pattern `b: Test.o1.B`, but got O(2)
	at scala.Predef$.assert(Predef.scala:223)
```

The class `D` has two separate outer fields which are not statically guaranteed to be the same.

```
jardiff  '/tmp/Test$O$D.class'

+++ Test$O$D.class.asm
// class version 52.0 (52)
// access flags 0x21
public class Test$O$D extends Test$O$B implements Test$O$C {

  // access flags 0x9
  public static INNERCLASS Test$O Test O
  // access flags 0x1
  public INNERCLASS Test$O$B Test$O B
  // access flags 0x601
  public abstract INNERCLASS Test$O$C Test$O C
  // access flags 0x1
  public INNERCLASS Test$O$D Test$O D

  // access flags 0x1011
  public final synthetic LTest$O; $outer

  // access flags 0x1
  public <init>(LTest$O;)V
    // parameter final synthetic  $outer
    ALOAD 1
    IFNONNULL L0
    ACONST_NULL
    ATHROW
   L0
    ALOAD 0
    ALOAD 1
    PUTFIELD Test$O$D.$outer : LTest$O;
    ALOAD 0
    GETSTATIC Test$.MODULE$ : LTest$;
    INVOKEVIRTUAL Test$.o2 ()LTest$O;
    INVOKESPECIAL Test$O$B.<init> (LTest$O;)V
    ALOAD 0
    INVOKESTATIC Test$O$C.$init$ (LTest$O$C;)V (itf)
    RETURN
    MAXSTACK = 2
    MAXLOCALS = 2

  // access flags 0x1001
  public synthetic Test$O$C$$$outer()LTest$O;
    ALOAD 0
    INVOKEVIRTUAL Test$O$D.Test$O$D$$$outer ()LTest$O;
    ARETURN
    MAXSTACK = 1
    MAXLOCALS = 1

  // access flags 0x1001
  public synthetic Test$O$D$$$outer()LTest$O;
    ALOAD 0
    GETFIELD Test$O$D.$outer : LTest$O;
    ARETURN
    MAXSTACK = 1
    MAXLOCALS = 1

  // access flags 0x1
  public cOuter()LTest$O;
    ALOAD 0
    INVOKESTATIC Test$O$C.cOuter$ (LTest$O$C;)LTest$O; (itf)
    ARETURN
    MAXSTACK = 1
    MAXLOCALS = 1
}
```

The existing logic in `outerTest` incorrectly assumes that the runtime check of the outer reference for the pattern `b: Test.o1.B` can be elided given that the scrutinee of the pattern `a: Test.o1.C` has the prefix `Test.o1`.

retronym committed:
More accurate outer checks in patterns

Avoids eliding outer checks that matter (run/t11534b.scala) and
avoids emitting checks that don't (pos/t11534.scala) which avoids
compiler warnings when the tested class doesn't have an outer
field.

The latter stops the annoying unchecked warning that appeared since
a recent refactoring made `TermName` a final class.

retronym commented:
I've pushed a new version of my change that seems a few paces closer to a final solution here.

joroKr21 commented:
> The class D has two separate outer fields which are not statically guaranteed to be the same.

Vague thought - what about using `baseType` to get the prefix wrt to the scrutinee?

retronym commented:
> Vague thought - what about using baseType to get the prefix wrt to the scrutinee?

Neither `B` nor `C` is a subclass of the other.

retronym commented:
/synch

joroKr21 commented:
> Neither B nor C is a subclass of the other.

In that case the conclusion could be no - we can't elide the check.

retronym commented:
/synch

joroKr21 commented:
NVM I see that you already use `baseType` - I should have checked. However I can't make any sense of `prefixAlign2`.

retronym commented:
> However I can't make any sense of prefixAlign2

Sorry, more docs and a good set of tests are still needed here! The idea is as follows.

Given:

`(x: a.B) match { case _: a.C }`

Where `C` is subclass of `A`

Is it possible that a value conforms to `a.B` and `a1.C` where `a ne a1`?

To answer this, I create a new prefix based on a fresh symbol and check the base type of `TypeRef(freshPrefix, typePatternSymbol, args)` at the binder symbol. If that is prefixed by the fresh symbol, they are statically the same.

dwijnand commented:
> @dwijnand Formulating the test this way might make my point a bit clearer:

Thanks.  I tried to disprove the assertion with

```scala
  trait BI extends o2.B
  class D2 extends o1.B with o1.C with BI

  // expected O(1) as outer of value conforming to pattern `b: o1.B`, but got O(2)
  def main(args: Array[String]): Unit = {
    pat1(new o1.D)
    pat1(new D2)
  }
```

but it won't let you define such a D2:

```
outer-checks.scala:16: error: illegal inheritance;
 class D2 inherits different type instances of class B:
Test.o2.B and Test.o1.B
  class D2 extends o1.B with o1.C with BI
        ^
```

which I _think_ I might've seen recently pop up as not being caught in Scala 3, but I might be confusing with something similar looking.

I'll try to review the patch.

dwijnand commented on src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala:
```diff
@@ -361,20 +358,78 @@ trait MatchTreeMaking extends MatchCodeGen with Debugging {
+
+          expectedTp match {
+            case _ if expectedTp.typeSymbolDirect.isAliasType =>
+              withOuterTest(orig)(testedBinder, expectedTp.dealias)
```
Are the usages of `expectedTp` fine with no longer using the original alias?

retronym replied:
This dealias is needed to have the behaviour in the cases below.

```
class Outer { class A; class B; type T = A with B }
val o = new Outer
type T1 = o.A with o.B
def pat1(a: Any) = a match { case o.T => }
def pat2(a: Any) = a match { case T1 => }
def pat3(a: Any) = a match { case o.A with o.A => }
```

I can't see any assumptions in the other usages of `expectedTp` that would be violated by recursing into the alias.

dwijnand commented on test/files/neg/t7721.check:
```diff
@@ -22,6 +22,24 @@ t7721.scala:48: warning: abstract type pattern B.this.Foo is unchecked since it
+         ^
+t7721.scala:48: warning: The outer reference in this type test cannot be checked at run time.
+    case x: Foo with Bar with Concrete => x.bippy + x.barry + x.dingo + x.conco + x.bongo
+         ^
```
Are these unfortunate casualties or is this progress?

retronym replied:
Previously refinement types had no outer check (even if all the components had the same prefix). Here's why:

```
cala> class Outer { class A; class B; type T = A with B }
class Outer

scala> :power
Power mode enabled. :phase is at typer.
import scala.tools.nsc._, intp.global._, definitions._
Try :help or completions for vals._ and power._

scala> val o = new Outer
val o: Outer = Outer@51044708

scala> val t1 = typeOf[o.T].dealias
val t1: $r.intp.global.Type = o.A with o.B

scala> t1.prefix
val res1: $r.intp.global.Type = <notype>

scala> t1.asInstanceOf[RefinedType].parents.map(_.prefix)
val res8: List[$r.intp.global.Type] = List(o.type, o.type)
```

This patch recurses into refinement types and adds outer tests for each element.

So I'd classify this as progress, although some code may have come to rely on the the weaker existing behaviour.

dwijnand reviewed (approved):
LGTM, for its results rather than large sections of how it's done.  I understand the intent though.  Nice tests and nice fix.  I left a couple of questions.

retronym commented:
@adriaanm Perhaps you'd also like to take a look at where this ended up.

retronym commented on src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala:
```diff
@@ -361,20 +358,78 @@ trait MatchTreeMaking extends MatchCodeGen with Debugging {
+
+          expectedTp match {
+            case _ if expectedTp.typeSymbolDirect.isAliasType =>
+              withOuterTest(orig)(testedBinder, expectedTp.dealias)
```
I'm not sure now why I didn't just write `expectedType.dealias match {` here without this case.

dwijnand replied:
From local tests that seems to have fixed the infinite loop issue I was seeing.  Fixed up the commit in my forward port PR.

dwijnand commented:
Likely going to avoid the risk in 2.12 and land #9504 instead, for 2.13.
